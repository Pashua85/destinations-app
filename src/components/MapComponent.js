import React from 'react';
import { compose, withProps } from 'recompose';
import { 
  withGoogleMap, 
  GoogleMap, 
  Marker, 
  Polyline, 
  InfoWindow
} from 'react-google-maps';

class Map extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      showInfoBox: false,
      selectAddress: null    
    };
  };

  componentDidUpdate() {
    console.log(this.state.pathCoords);
    console.log(this.props.addresses);
    if(this.props.addresses.length > 1) {
      const bounds = new window.google.maps.LatLngBounds();
      this.props.addresses.map((address, i) => {
        bounds.extend(new window.google.maps.LatLng(
          address.latLng.lat,
          address.latLng.lng
        ));
      });
      this.map.fitBounds(bounds);
    }
  }

  handleClickedMarker = (address) => () => {
    this.setState(() => ({
      selectAddress: address,
      showInfoBox: true
    }))
  };

  onStartChangePosition = ()  => {
    this.setState(() => ({
      showInfoBox: false
    }))
  };

  render() {    
    var center;
    this.props.addresses.length ?
    center = this.props.addresses[0].latLng :
    center = {
      lat: 56.80950300000001,
      lng: 60.606506999999965
    };

    return (
      <GoogleMap
        ref={ref => { this.map = ref;}}
        defaultZoom={12}  
        center={center}
      > 
        { this.props.addresses &&
        <Polyline
          path={this.props.pathCoords}
          options={{
            strokeColor: '#FF0000',
            strokeOpacity: 1.0
          }}
        />
        }
        { 
          this.props.addresses &&
          this.props.addresses.map((address, index) => {
            return (
              <Marker 
                key={address.placeId}
                position={address.latLng} 
                draggable={true}
                onDragStart={this.onStartChangePosition}
                onDragEnd={(event) => this.props.onChangePosition(address.placeId, event.latLng)}
                onClick={this.handleClickedMarker(address)}
              />
            )
          })
        }   
        {
          this.state.showInfoBox && <InfoWindow
            key={Math.random()}
            onClick={this.handleCloseInfoBox}
            position={this.state.selectAddress.latLng}
          >
            <p>{this.state.selectAddress.formattedAddress}</p>
          </InfoWindow>
        }
      </GoogleMap>
    )
  }
};

const MyMapComponent = compose(
  withProps({
    loadingElement: <div style={{ height: `100%` }} />,
    containerElement: <div style={{ height: `400px` }} />,
    mapElement: <div style={{ height: `100%` }} />,
  }),
  withGoogleMap
)(Map);

export default MyMapComponent;