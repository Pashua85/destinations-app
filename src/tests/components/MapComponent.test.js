import React from 'react';
import { shallow } from 'enzyme';
import MapComponent from '../../components/MapComponent';
import addresses from '../helpers/addresses';
import pathCoords from '../helpers/pathCoords';


test('should render MapComponent correctly', () => {
  const wrapper = shallow(
    <MapComponent addresses={addresses} pathCoords={pathCoords} />
  );
  expect(wrapper).toMatchSnapshot();
});