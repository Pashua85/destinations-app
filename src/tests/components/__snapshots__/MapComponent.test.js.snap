// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render MapComponent correctly 1`] = `
<withGoogleMap(Map)
  addresses={
    Array [
      Object {
        "formattedAddress": "ул. Щорса, 94А, Екатеринбург, Свердловская обл., Россия, 620144",
        "latLng": Object {
          "lat": 56.80950300000001,
          "lng": 60.606506999999965,
        },
        "placeId": "ChIJc8sXNttuwUMRWSTSqEBCxrY",
      },
      Object {
        "formattedAddress": "ул. Чайковского, 62, Екатеринбург, Свердловская обл., Россия, 620130",
        "latLng": Object {
          "lat": 56.8073729,
          "lng": 60.62031009999998,
        },
        "placeId": "ChIJASguT89uwUMRMbx0H8V2uNk",
      },
      Object {
        "formattedAddress": "ул. Студенческая, 3, Екатеринбург, Свердловская обл., Россия, 620137",
        "latLng": Object {
          "lat": 56.854948,
          "lng": 60.661017000000015,
        },
        "placeId": "ChIJ6TiVJjVswUMRRi8Evo28BM4",
      },
      Object {
        "formattedAddress": "Москва, Россия",
        "latLng": Object {
          "lat": 55.755826,
          "lng": 37.617299900000035,
        },
        "placeId": "\\"ChIJybDUc_xKtUYRTM9XV8zWRD0\\"",
      },
    ]
  }
  className="map-here"
  containerElement={
    <div
      style={
        Object {
          "border": "1px solid green",
          "height": "400px",
          "width": "66%",
        }
      }
    />
  }
  loadingElement={
    <div
      style={
        Object {
          "height": "100%",
        }
      }
    />
  }
  mapElement={
    <div
      style={
        Object {
          "height": "100%",
        }
      }
    />
  }
  pathCoords={
    Array [
      Object {
        "lat": 56.80950300000001,
        "lng": 60.606506999999965,
      },
      Object {
        "lat": 56.8073729,
        "lng": 60.62031009999998,
      },
      Object {
        "lat": 56.854948,
        "lng": 60.661017000000015,
      },
      Object {
        "lat": 55.755826,
        "lng": 37.617299900000035,
      },
    ]
  }
/>
`;
